blog/content/post/2025-04-18-cache-attack-mitigation-testing.md

```diff
@@ -0,0 +1,350 @@
+placed into one, evicting occupied one if needed. Which specific way is chosen
+depends on the replacement policy.
+
+There are various implementations of this policy, such as LRU (Least Recently
+Used), random, etc. The replacement policy depends on the architecture and, if
+multiple options are available, on the system configuration. One example is the
+Cortex-A72, where the L2 replacement policy can be set to either PLRU
+(pseudo-LRU) or pseudo-random.
+
+## Shared memory
+
+A lot of cache timing attacks works by using shared memory. It works by gaining
+information which shared address was used by program or process. Those types of
+attacks are usually simpler to implement as we can know exactly what data and
+where is it stored.
+
+Example of memory shared by 2 processes.
+
+![shared memory virtual address to physical address](/img/shared-mem.png)
+
+Such memory can hold various types of information e.g. executable library code
+(e.g. functions) or constant global data. While these might seem like
+non-sensitive information, knowing who accessed which data and in which order
+can reveal much more confidential information.
+
+Some implementations of AES are good example of this. They use lookup tables to
+speed up encryption and decryption. The tables themselves are not secret, but
+gaining knowledge about which elements were accessed can lead to recovering
+encryption key. This is because access to different array elements can depend
+not only on e.g. text to be encrypted but the key itself.
+
+## Cache timing attacks
+
+Cache timing attacks are a type of side-channel attack, which means they rely on
+indirect information such as power consumption, memory access time, etc.
+
+In the case of cache timing attacks, the attacker manipulates the cache (either
+directly through available instructions or indirectly by exploiting how the
+cache works) and measures timing of: memory access, different cache operations,
+program execution time or similar.
+
+Most of these attacks differ in their granularity. The most precise attacks can
+determine exactly which block or line was accessed, less precise can only
+determine which cache set was used.
+
+### Attack types
+
+Below is description of 3 different attack types which I tried to implement:
+
+* `Flush+Reload` - appears to be one of the simpler attacks. It consists of 2
+  parts:
+  * Flush: evicting shared address from the cache
+  * Reload: measuring evicted memory access time
+
+  If the timing measurement indicates a cache hit, it means that someone
+  accessed the given address between Flush and Reload. To improve detection
+  (and reduce false negatives), a short wait is often added between Flush and
+  Reload. This attack allows information to be obtained with cache line
+  granularity.
+
+  The image below[^flush+reload] illustrates the basic mechanism of this
+  attack:
+
+  ![Flush+Reload timing](/img/flush_reload_timing.png)
```

Comment:
The description under the picture is not good, it has formatting issues (much bigger font), and I don't think it is needed because it does not add much value. I would appreciate it if you would adjust the cutting.